Respect controllers on PVCs for retention policy

mattcary opened the pull request:
/kind bug

Fixes #122400

/assign @msau42 

```release-note
StatefulSet autodelete will respect controlling owners on PVC claims as described in https://github.com/kubernetes/enhancements/pull/4375
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

- [KEP #1847](https://github.com/kubernetes/enhancements/issues/1847)

mattcary committed:
Respect controllers on PVCs for retention policy

mattcary commented:
/sig apps
/priority-soon

mattcary commented:
/triage accepted

mattcary commented:
/priority important-soon

msau42 commented:
/lgtm
/approve

carlory commented:
/assign @soltysh

atiratree reviewed (commented):
@mattcary I found couple of problems with the current implementation

I feel that `hasOwnerRef` should have a bool `checkController` parameter. And we should not check the controller flag if we want to delete the owner ref, but we should check it if we want to introduce it / overwrite it.

I think this goes for all the cases in both `claimOwnerMatchesSetAndPod` and `updateClaimOwnerRefForSetAndPod` functions.

\+ would be good to document why we are doing it

atiratree commented on pkg/controller/statefulset/stateful_pod_control.go:
```diff
@@ -241,6 +241,10 @@ func (spc *StatefulPodControl) UpdatePodClaimForRetentionPolicy(ctx context.Cont
+		case hasUnexpectedController(claim, set, pod):
+			// Do not set any ownerRefs if there is an unexpected controller.
+			msg := fmt.Sprintf("Claim %s has a conflicting controller, the retention policy is ignored for this claim", claimName)
+			spc.recorder.Event(set, v1.EventTypeWarning, "ConflictingController", msg)
```
also this part from the KEP seems not to be satisfied and should apply for all policy combinations: 

> If there is another controller, the existing non-controller StatefulSet or Pod reference will be removed.

mattcary replied:
Thanks for taking a look at this! You make some good points.

Let's resolve the conversation on the KEP first about what the proper semantics should be, then I'll work through your comments here -- you raise some excellent points.

mattcary replied:
argh, you are correct. Thank you.

mattcary replied:
Change is in progress. It was better to make a little more complete re-vamp of the logic to keep it from getting too complicated.

atiratree replied:
Thanks for looking at that. And yeah, there are a lot of cases that need to be covered. It would be nice to have them all covered by the tests.

mattcary commented on pkg/controller/statefulset/stateful_set_utils.go:
```diff
@@ -305,6 +333,7 @@ func setOwnerRef(target, owner metav1.Object, ownerType *metav1.TypeMeta) bool {
 			Kind:       ownerType.Kind,
 			Name:       owner.GetName(),
 			UID:        owner.GetUID(),
+			Controller: ptr.To(true),
```
hasOwnerRef checks that there is not an owner with the same UID, regardless of the value of the controller.

Although, it occurs to me, if the StatefulSet is updated, the UID will change. So we probably want hasOwnerRef to check on api version + kind + name + namespace, which will be unique?

mattcary replied:
Ah I didn't know the ownerref was removed with the orphan strategy.

Mostly I think I got confused.

I guess the problem could be a pvc that's slow to delete, and a statefulset that's deleted and recreated (the same case the stale pod claim check deals with). In that case though a new pod won't bind to a deleted PVC, and so it should eventually delete and the statefulset controller will recreate a new PVC.

So yeah, we should just look at UIDs (except for the stale pod case).

mattcary commented on pkg/controller/statefulset/stateful_set_utils.go:
```diff
@@ -305,6 +333,7 @@ func setOwnerRef(target, owner metav1.Object, ownerType *metav1.TypeMeta) bool {
+			Controller: ptr.To(true),
 		})
 	target.SetOwnerReferences(ownerRefs)
 	return true
```
+1 this will be fixed

mattcary commented:
/ok-to-test

mattcary commented:
@atiratree it's ready to be reviewed now, thanks for you patience.

I changed updateClaimOwnerRefForSetAndPod to be more of a reconcilliation style (clear our refs out to a known state, then set them how they should be). This seemed to be the best way to deal with all the possible corner cases.  I think it addresses all your concerns.

mattcary commented:
@atiratree, any chance you can take another look at this? Thanks!

atiratree commented on pkg/controller/statefulset/stateful_set_utils_test.go:
```diff
@@ -417,71 +616,623 @@ func TestUpdateClaimOwnerRefForSetAndPod(t *testing.T) {
+		claim := v1.PersistentVolumeClaim{}
+		claim.SetOwnerReferences(tc.originalRefs)
+		updateClaimOwnerRefForSetAndPod(logger, &claim, &set, &pod)
+		if ownerRefsChanged(tc.expectedRefs, claim.GetOwnerReferences()) {
```
can we simplify this and just do `DeepEqual`?

mattcary replied:
DeepEqual won't work if the order of the references changes?

Being order-independent will make the test less fragile and less of a change-detector.

atiratree replied:
Thanks

atiratree commented on pkg/controller/statefulset/stateful_set_utils.go:
```diff
@@ -214,64 +291,54 @@ func claimOwnerMatchesSetAndPod(logger klog.Logger, claim *v1.PersistentVolumeCl
 	case policy.WhenScaled == retain && policy.WhenDeleted == delete:
-		needsUpdate = setOwnerRef(claim, set, &setMeta) || needsUpdate
-		needsUpdate = removeOwnerRef(claim, pod) || needsUpdate
+		newRefs = addControllerRef(newRefs, set, controllerKind)
```
+1 nice refactoring

mattcary replied:
thx

atiratree commented on pkg/controller/statefulset/stateful_pod_control_test.go:
```diff
@@ -552,30 +553,53 @@ func TestStatefulPodControlUpdatePodClaimForRetentionPolicy(t *testing.T) {
+				t.Errorf("Could not add claim %s: %v", k, err)
+			}
+		}
+		pod1 := newStatefulSetPod(set, 1)
```
This is basically two tests in one. Can we parametrize this?

mattcary replied:
done

atiratree commented on test/e2e/apps/statefulset.go:
```diff
@@ -1339,6 +1340,69 @@ var _ = SIGDescribe("StatefulSet", func() {
+					Name:       "dummy-controller",
+					APIVersion: "dummy/v1",
+					Kind:       "dummy",
+					UID:        "123",
```
can we generate real fake UUID for e2e tests?

mattcary replied:
done

atiratree commented on test/e2e/apps/statefulset.go:
```diff
@@ -1400,6 +1464,73 @@ var _ = SIGDescribe("StatefulSet", func() {
+				APIVersion: "unknown/v1",
+				Kind:       "controller",
+				Name:       "unknown",
+				UID:        "unknown",
```
and UUID here

mattcary replied:
done

atiratree commented on test/e2e/apps/statefulset.go:
```diff
@@ -1400,6 +1464,73 @@ var _ = SIGDescribe("StatefulSet", func() {
+					}
+					found := false
+					for _, ref := range claim.GetOwnerReferences() {
+						if ref.Kind == "StatefulSet" && ref.Name == ssName {
```
can we check the full shape of the owner reference with DeepEqual?

mattcary replied:
Done.

atiratree commented on test/e2e/apps/statefulset.go:
```diff
@@ -1400,6 +1464,73 @@ var _ = SIGDescribe("StatefulSet", func() {
+				}
+				return true, nil // found them both!
+			})
+			framework.ExpectNoError(err)
```
can we remove the controller flag from one of the claims here and check it gets reconciled back to true?

mattcary replied:
Done

atiratree commented on test/e2e/apps/statefulset.go:
```diff
@@ -1400,6 +1464,73 @@ var _ = SIGDescribe("StatefulSet", func() {
+			ginkgo.By("Add external controller to PVC 1")
+			claim, err := c.CoreV1().PersistentVolumeClaims(ns).Get(ctx, claimNames[1], metav1.GetOptions{})
+			framework.ExpectNoError(err)
+			claim.SetOwnerReferences(append(claim.GetOwnerReferences(), metav1.OwnerReference{
```
might be worth to also add non existing StatefulSet to one of the PVC and see it does not get deleted as well.

mattcary replied:
done

atiratree reviewed (commented):
> I changed updateClaimOwnerRefForSetAndPod to be more of a reconcilliation style (clear our refs out to a known state, then set them how they should be). This seemed to be the best way to deal with all the possible corner cases. I think it addresses all your concerns.

@mattcary yes, looks much better now. But needs some polishing.

mattcary commented on pkg/controller/statefulset/stateful_set_utils_test.go:
```diff
@@ -417,71 +616,623 @@ func TestUpdateClaimOwnerRefForSetAndPod(t *testing.T) {
+		pod.SetUID("pod-uid")
+		set.Spec.PersistentVolumeClaimRetentionPolicy = nil
+		if hasUnexpectedController(target, set, pod) {
+			t.Errorf("Got unexpected controller at %s when no retention policy", tc.name)
```
done

mattcary commented on pkg/controller/statefulset/stateful_set_utils_test.go:
```diff
@@ -417,71 +616,623 @@ func TestUpdateClaimOwnerRefForSetAndPod(t *testing.T) {
+			set.Spec.PersistentVolumeClaimRetentionPolicy = &policy
+			got := hasUnexpectedController(target, set, pod)
+			if got != tc.expected {
+				t.Errorf("Unexpected other controller at %s (policy %v)", tc.name, policy)
```
done

mattcary commented on pkg/controller/statefulset/stateful_set_utils_test.go:
```diff
@@ -417,71 +616,623 @@ func TestUpdateClaimOwnerRefForSetAndPod(t *testing.T) {
+	testCases := []struct {
+		name     string
+		refs     []metav1.OwnerReference
+		expected bool
```
done

mattcary commented on pkg/controller/statefulset/stateful_set_utils_test.go:
```diff
@@ -417,71 +616,623 @@ func TestUpdateClaimOwnerRefForSetAndPod(t *testing.T) {
+					Kind:       "CustomerController",
+					Name:       "simon",
+					UID:        "other-uid",
+					Controller: ptr.To(true),
```
seems cheap to leave it in though --- whatever validation is done is not very well documented at the least.

mattcary replied:
Done

mattcary commented on pkg/controller/statefulset/stateful_pod_control.go:
```diff
@@ -242,14 +242,16 @@ func (spc *StatefulPodControl) UpdatePodClaimForRetentionPolicy(ctx context.Cont
 		default:
+			if hasUnexpectedController(claim, set, pod) {
+				// Add an event so the user knows they're in a strange configuration. The claim will be cleaned up below.
+				msg := fmt.Sprintf("Claim %s has a conflicting controller, the retention policy is ignored for this claim", claimName)
```
done

mattcary commented on pkg/controller/statefulset/stateful_set_utils_test.go:
```diff
@@ -417,71 +616,623 @@ func TestUpdateClaimOwnerRefForSetAndPod(t *testing.T) {
+					Kind:       "CustomerController",
+					Name:       "simon",
+					UID:        "other-uid",
+					Controller: ptr.To(false),
```
ditto

mattcary commented on pkg/controller/statefulset/stateful_set_utils_test.go:
```diff
@@ -417,71 +616,623 @@ func TestUpdateClaimOwnerRefForSetAndPod(t *testing.T) {
+					APIVersion: "apps/v1",
+					Kind:       "StatefulSet",
+					Name:       "set",
+					UID:        "stale-uid",
```
Yeah, that seems to be a good point, thanks.

mattcary commented on pkg/controller/statefulset/stateful_set_utils.go:
```diff
@@ -170,9 +171,85 @@ func getPersistentVolumeClaimRetentionPolicy(set *apps.StatefulSet) apps.Statefu
+		return false
+	}
+	for _, ownerRef := range claim.GetOwnerReferences() {
+		if matchesRef(&ownerRef, set, controllerKind) || matchesRef(&ownerRef, pod, podKind) {
```
Yeah, I agree, I've done that for both the pod & the set, and added a comment to capture my understanding. LMK if the comment  shows I've misunderstood.

mattcary commented on pkg/controller/statefulset/stateful_set_utils_test.go:
```diff
@@ -417,71 +616,623 @@ func TestUpdateClaimOwnerRefForSetAndPod(t *testing.T) {
+					Kind:       "Set",
+					Name:       "set",
+					UID:        "set",
+					Controller: ptr.To(true),
```
I'm not sure why admission validation affects the behavior we want to test for. This way we know we have consistent behavior even if something goes awry with validation.

+1 on testing variations on controller vs not controller.

mattcary replied:
done

mattcary commented on pkg/controller/statefulset/stateful_set_utils.go:
```diff
@@ -170,9 +171,85 @@ func getPersistentVolumeClaimRetentionPolicy(set *apps.StatefulSet) apps.Statefu
+}
+
+// nonController returns true if the pod or set is an owner but not controller of the claim.
+func nonController(claim *v1.PersistentVolumeClaim, set *apps.StatefulSet, pod *v1.Pod) bool {
```
done

mattcary commented on pkg/controller/statefulset/stateful_set_utils.go:
```diff
@@ -170,9 +171,85 @@ func getPersistentVolumeClaimRetentionPolicy(set *apps.StatefulSet) apps.Statefu
+// nonController returns true if the pod or set is an owner but not controller of the claim.
+func nonController(claim *v1.PersistentVolumeClaim, set *apps.StatefulSet, pod *v1.Pod) bool {
+	for _, ownerRef := range claim.GetOwnerReferences() {
+		if ownerRef.UID == set.GetUID() || ownerRef.UID == pod.GetUID() {
```
I'm not sure what you mean --- both set.GetUID & opd.GetUID are checked?

mattcary replied:
This is checking that there is an owner that is not a controller -- a non-controller, as the function comment explains.

mattcary commented on pkg/controller/statefulset/stateful_set_utils.go:
```diff
@@ -214,64 +291,54 @@ func claimOwnerMatchesSetAndPod(logger klog.Logger, claim *v1.PersistentVolumeCl
+	newRefs := make([]metav1.OwnerReference, len(originalRefs))
+	copy(newRefs, originalRefs)
+
+	unexpectedController := hasUnexpectedController(claim, set, pod)
```
What I found when refactoring this is that it gets very confusing & hard to maintain if we try to combine update logic and validation logic.

I think it will be a lot more understandable and reliable to have the update code just update and the validation code (isClaimOwnerUpToDate) just validate. Otherwise we can get into many weird corner cases and hard to reason about code.

mattcary replied:
It's not necessary to call isClaimOwnerUpToDate first. After calling updateClaimOwnerRefForSetAndPod, usually the caller will make an api server update call which is potentially expensive, so calling isClaimOwnerUpToDate avoids that.

That seems to be the logic in stateful_set_control.go, eg see around line 442.

mattcary commented on pkg/controller/statefulset/stateful_set_utils.go:
```diff
@@ -170,9 +171,85 @@ func getPersistentVolumeClaimRetentionPolicy(set *apps.StatefulSet) apps.Statefu
+
+// removeRefs removes any owner refs from the list matching predicate. Returns true if the list was changed and
+// the new (or unchanged list).
+func removeRefs(refs []metav1.OwnerReference, predicate func(ref *metav1.OwnerReference) bool) (bool, []metav1.OwnerReference) {
```
+1 good observation thx. That was left over from an earlier version -- but it got out of hand to try to track changes to resources which is why I went to this more clear isClaimOwnerUpToDate / updateClaimOwnerRef separation.

mattcary commented on pkg/controller/statefulset/stateful_set_utils.go:
```diff
@@ -214,64 +291,54 @@ func claimOwnerMatchesSetAndPod(logger klog.Logger, claim *v1.PersistentVolumeCl
+func updateClaimOwnerRefForSetAndPod(logger klog.Logger, claim *v1.PersistentVolumeClaim, set *apps.StatefulSet, pod *v1.Pod) {
+	originalRefs := claim.GetOwnerReferences()
+	newRefs := make([]metav1.OwnerReference, len(originalRefs))
+	copy(newRefs, originalRefs)
```
yes. We also unconditionally set the refs at the end anyway. thx.

mattcary commented:
Thanks for the review! 

The push I just did contains all the responses to the comments --- I did them all, except the ones around validation + a couple more comments/questions.

The e2e test may not be passing, I'll check into that tomorrow.

mattcary commented:
The e2e test passed locally.

Note that the unknown controller ref still causes the pvc to be deleted. But that's on the controller semantics, not us; I verify in the test we don't update the controller.

mattcary commented:
Hmm, the e2e test are passing locally, but not in the CI. Investigating.

mattcary commented:
Hmm, something was strange in my e2e test logic. fiddling.

mattcary commented:
Ok, cleaned it up. I think the problem was accidentally looking for the unknown claim, along with a race condition on finding or not finding the first claim. It's passing locally, let's see what happens in the CI.

mattcary commented:
Uff, that was tricky.

I didn't realize that the stale owner ref would trigger the garbage collector on the PVC. When filtering the e2e run to just the single test, the GC doesn't get called, but when running in the presubmit it gets called much more frequently. So the behavior was different locally than in the presubmit.

I used a finalizer to pin the PVC with the stale ref so we can test the update behavior.

If the tests pass this should be good to review.

mattcary commented:
/retest pull-kubernetes-conformance-kind-ipv6-parallel 

Flake in the n/w test.

mattcary commented:
/test pull-kubernetes-conformance-kind-ipv6-parallel

mattcary commented:
fyi since the 1.30 freeze has started, I'll wait until the thaw next month this, then submit this bug fix & backport to 1.30.

atiratree commented on test/e2e/apps/statefulset.go:
```diff
@@ -1400,6 +1465,175 @@ var _ = SIGDescribe("StatefulSet", func() {
+			ginkgo.By("Delete the stateful set and wait for claims 0 and 2 but not 1 and 3 to disappear")
+			err = c.AppsV1().StatefulSets(ns).Delete(ctx, ssName, metav1.DeleteOptions{})
+			framework.ExpectNoError(err)
+			err = verifyStatefulSetPVCsExist(ctx, c, ss, []int{1, 3})
```
how come 1 is still there if the expectedExternalRef doesn't have the controller flag? Am I missing something?

mattcary replied:
An external owner ref is a stale owner ref, see stateful_set_utils.go:isClainOwnerUpToDate.

atiratree replied:
I understand it better now. It reconciles normally and still has both refs: `slices.Concat(expectedExternalRef, expectedOwnerRef)`

It does not get deleted because of the first one. Unfortunately because we create a non-existent reference we fire the following errors in the KCM garbage collector:

```
E0501 00:27:21.573617  154898 garbagecollector.go:397] "error syncing item" err="unable to get REST mapping for unknown/v1/randomOwner." logger="garbage-collector-controller" item="[v1/PersistentVolumeClaim, namespace: statefulset-7724, name: datadir-ss-1, uid: 6f062806-1064-4df6-a6da-63c2cd3d7929]"
```

which I think is rather unfortunate, and it would be better if we could prevent that by creating an existing reference or at least an existing kind.

atiratree replied:
+1. Can we also properly comment out why the PVCs are not getting deleted at the end?

mattcary replied:
done

atiratree reviewed (commented):
GJ on rewriting the code and the tests. I think the readability is good, but unfortunately it still takes a bit of time to understand the code, but this is due to the problem domain...

mattcary commented on pkg/controller/statefulset/stateful_set_utils_test.go:
```diff
@@ -417,71 +616,623 @@ func TestUpdateClaimOwnerRefForSetAndPod(t *testing.T) {
+					APIVersion: "v1",
+					Kind:       "Pod",
+					Name:       "pod",
+					UID:        "stale-uid",
```
Yes, this test case confirms that we report that as unexpected, as above?

mattcary commented on test/e2e/apps/statefulset.go:
```diff
@@ -1400,6 +1465,175 @@ var _ = SIGDescribe("StatefulSet", func() {
+					BlockOwnerDeletion: ptr.To(true),
+				},
+			}
+			ginkgo.By("Expect claims 0 and 2 to have ownerRefs to the statefulset, and 3 to have a stale reference")
```
done

mattcary commented on test/e2e/apps/statefulset.go:
```diff
@@ -1400,6 +1465,175 @@ var _ = SIGDescribe("StatefulSet", func() {
+					APIVersion: "unknown/v1",
+					Kind:       "controller",
+					Name:       "unknown",
+					UID:        "04a46669-0625-4466-8614-70412988bf19",
```
This is supposed to be an external owner ref, not a controller. I've clarified the name.

mattcary commented on pkg/controller/statefulset/stateful_set_utils_test.go:
```diff
@@ -417,71 +704,651 @@ func TestUpdateClaimOwnerRefForSetAndPod(t *testing.T) {
+			pod.SetUID("pod-456")
+			claim := v1.PersistentVolumeClaim{}
+			claimRefs := claim.GetOwnerReferences()
+			if hasPodRef {
```
Done, further down.

mattcary commented on pkg/controller/statefulset/stateful_set_utils_test.go:
```diff
@@ -360,14 +467,106 @@ func TestClaimOwnerMatchesSetAndPod(t *testing.T) {
+					APIVersion: "v1",
+					Kind:       "Pod",
+					UID:        "pod-123",
+					Controller: ptr.To(true),
```
The test case name serves as the comment, I've clarified it.

mattcary commented on pkg/controller/statefulset/stateful_set_utils.go:
```diff
@@ -170,9 +171,99 @@ func getPersistentVolumeClaimRetentionPolicy(set *apps.StatefulSet) apps.Statefu
+	}
+
+	if hasNonController(claim, set, pod) {
+		return false
```
done

mattcary commented:
@atiratree , I've responded and/or implemented all of your comments as far as I can tell, let me know what you think.

Thanks for your review!

atiratree reviewed (commented):
@mattcary there are still some nits and testing considerations

atiratree commented on test/e2e/apps/statefulset.go:
```diff
@@ -1400,6 +1479,196 @@ var _ = SIGDescribe("StatefulSet", func() {
+			ginkgo.By("Create configmap to use as random owner")
+			randomConfigMap, err := c.CoreV1().ConfigMaps(ns).Create(ctx, &v1.ConfigMap{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "random-owner",
```
nit: this could be a `generateName`

mattcary replied:
Done, good idea.

atiratree reviewed (commented):
There are 2 nits, but the PR lgtm.

Nevertheless, it would be good to get another set of eyes on it. 

cc @soltysh

atiratree commented:
unit test flaking due to https://github.com/kubernetes/kubernetes/issues/124526
/test pull-kubernetes-unit

soltysh commented on test/e2e/apps/statefulset.go:
```diff
@@ -1339,6 +1342,82 @@ var _ = SIGDescribe("StatefulSet", func() {
+
+			ginkgo.By("Remove PVC 1 owner ref")
+			pvc.OwnerReferences = nil
+			_, err = c.CoreV1().PersistentVolumeClaims(ns).Update(ctx, pvc, metav1.UpdateOptions{})
```
I've tried running this test locally and the first two invocations I hit:
```
  [FAILED] Operation cannot be fulfilled on persistentvolumeclaims "datadir-ss-1": the object has been modified; please apply your changes to the latest version and try again
  In [It] at: k8s.io/kubernetes/test/e2e/apps/statefulset.go:1402 @ 06/04/24 12:13:52.29
```
so I'd suggest to go through all the PVCs update operations and make them retriable. Similarly StatefulSet updates should be changed to use `updateStatefrulSetWithRetries` most likely.

mattcary replied:
Done.

Interesting, I've been using kind and never hit that. But +1 to fewer flakes.

soltysh commented on pkg/controller/statefulset/stateful_set_utils.go:
```diff
@@ -170,9 +171,100 @@ func getPersistentVolumeClaimRetentionPolicy(set *apps.StatefulSet) apps.Statefu
-// claimOwnerMatchesSetAndPod returns false if the ownerRefs of the claim are not set consistently with the
+// matchesRef returns true when the object matches the owner reference, that is the name and GVK are the same.
+func matchesRef(ref *metav1.OwnerReference, obj metav1.Object, gvk schema.GroupVersionKind) bool {
+	return gvk.GroupVersion().String() == ref.APIVersion && gvk.Kind == ref.Kind && ref.Name == obj.GetName()
```
Why not using `IsControlledBy` from `k8s.io/apimachinery/pkg/apis/meta/v1`

mattcary replied:
That matches on the UID, not on the GVK & name. So if, eg, the statefulset was updated, IsControlledBy won't return true until the UID is updated?

soltysh commented on pkg/controller/statefulset/stateful_set_utils.go:
```diff
@@ -170,9 +171,100 @@ func getPersistentVolumeClaimRetentionPolicy(set *apps.StatefulSet) apps.Statefu
+				return true
+			}
+			continue // This is us.
+		}
```
Why do we need to match both statefulset and pod in the loop above?

mattcary replied:
PVC are controlled by the statefulset for the WhenDeleted retention type, and controlled by the pod for the WhenScaled retention type.

soltysh commented on pkg/controller/statefulset/stateful_set_utils.go:
```diff
@@ -170,9 +171,100 @@ func getPersistentVolumeClaimRetentionPolicy(set *apps.StatefulSet) apps.Statefu
+		return true
+	}
+
+	if hasNonControllerOwner(claim, set, pod) {
```
I don't understand why we should react in such a situation? It doesn't matter that we don't own it, and furthermore who owns it. We just don't react on it.

soltysh replied:
At most we can warn the user about such situation.

mattcary replied:
The implementation before this change makes statefulsets & pods owners but not controllers -- that's what this PR is fixing. So we have to handle at least that case to correctly update existing statefulsets from before this bugfix.

soltysh reviewed (commented):
A few questions and suggestions to resolve  before the merge.

soltysh reviewed (approved):
/lgtm
/approve

mattcary commented:
/cherry-pick release-1.30

mattcary commented:
/cherry-pick release-1.29

mattcary commented:
/cherry-pick release-1.28
